src/Control/Effect/Class.hs

```diff
@@ -16,73 +16,87 @@ import GHC.Generics
 -- | The class of effect types, which must:
 --
 --   1. Be functorial in their last two arguments, and
---   2. Support threading effects in higher-order positions through using the carrier’s suspended state.
+--   2. Support threading effects in higher-order positions through using the carrier’s suspended context.
 --
--- All first-order effects (those without existential occurrences of @m@) admit a default definition of 'handle' provided a 'Generic1' instance is available for the effect.
+-- All first-order effects (those without existential occurrences of @m@) admit a default definition of 'thread' provided a 'Generic1' instance is available for the effect.
 --
 -- @since 1.0.0.0
 class Effect sig where
-  -- | Handle any effects in a signature by threading the carrier’s state all the way through to the continuation.
-  handle :: (Functor f, Monad m)
-         => f ()
-         -> (forall x . f (m x) -> n (f x))
-         -> sig m a
-         -> sig n (f a)
-  default handle :: (Functor f, Monad m, Generic1 (sig m), Generic1 (sig n), GEffect m n (Rep1 (sig m)) (Rep1 (sig n)))
-                 => f ()
-                 -> (forall x . f (m x) -> n (f x))
-                 -> sig m a
-                 -> sig n (f a)
-  handle state handler = to1 . ghandle state handler . from1
-  {-# INLINE handle #-}
+  -- | Handle any effects in a signature by threading the algebra’s handler all the way through to the continuation, starting from some initial context.
+  --
+  -- The handler is expressed as a /distributive law/, and required to adhere to the following laws:
+  --
+  -- @
+  -- handler . 'fmap' 'pure' = 'pure'
+  -- @
+  -- @
+  -- handler . 'fmap' (k '=<<') = handler . 'fmap' k 'Control.Monad.<=<' handler
+  -- @
+  --
+  -- respectively expressing that the handler does not alter the context of pure computations, and that the handler distributes over monadic composition.
+  thread
+    :: (Functor ctx, Monad m)
+    => ctx ()                              -- ^ The initial context.
+    -> (forall x . ctx (m x) -> n (ctx x)) -- ^ A handler for actions in a context, producing actions with a derived context.
+    -> sig m a                             -- ^ The effect to thread the handler through.
+    -> sig n (ctx a)
```

Comment:
As [has this](https://github.com/fused-effects/fused-effects/pull/296/files#r339126753), tho we haven’t added a `Monad` constraint on `n` in this PR.